### ISSUE TITLE
[Question] Search by attribute presence

### ISSUE THREAD

**Issue by wwwald** [zadam/trilium#1764]
Working with version 0.46.5.

Probably a stupid question, but I failed to find how to do this in the wiki or issues: how do I search for notes, that simply own a specific attribute?  

Concretely: I have a number of notes with an empty label `#note` assigned. What search query allows me to find these notes?

**Comment by zadam** [comment]
Search for `#note`, that's it ...

**Comment by wwwald** [comment]
OK, that was a terrible question statement :-)

My core problem is that I want this to work for inherited attributes.  
You know the scenario - but just to be sure: I have a template note `template-call` with the `#call` attribute, and several call notes inherit from that. In a search for `#project=myProject #call` I would like these individual call notes to show up. Currently in 0.46.5, I can't get this to work.

I've found more on this in #989 - seems to suggest that this is not a trivial feature to implement. #980 does suggest that this was integrated in the 0.43 search overhaul.  
Is searching on inherited attributes supposed to be working?

**Comment by zadam** [comment]
Yes, since 0.44 searching for e.g. `#call` will also return notes which inherit this attribute (through `isInheritable` or through a template). I'm not sure why this doesn't work for you.

When you search for `#call` do you also not get the results (to eliminate the influence of the `#project=myProject`)

**Comment by wwwald** [comment]
Searching by `#call` only _does_ return me all call notes.

Further experimenting: search string `#project=myProject #call` does not work, but `#call #project=myProject` does.

**Comment by zadam** [comment]
That is weird. When I simulate this scenario then both these variants work (as I understand they differ only by the filter order).

It would be helpful if you'd do the following steps:

1.  create full search note with the non-working search string
2.  add "Debug" search option
3.  Click on execute
4.  Click on global menu (top left), show backend log.

And post the debug output, like this:

![](https://user-images.githubusercontent.com/617641/112390454-a4d1fa00-8cf6-11eb-9cf7-c55fdd8be700.png)

**Comment by wwwald** [comment]
OK - glad to help out on this. See below - I only changed the actual project name to `myProject`.

Working search = `#call #project=myProject`
```
07:45:32.456 Fulltext tokens: []
07:45:32.456 Expression tokens: [
    {
        "token": "#call",
        "inQuotes": false,
        "startIndex": 0,
        "endIndex": 4
    },
    {
        "token": "#project",
        "inQuotes": false,
        "startIndex": 6,
        "endIndex": 13
    },
    {
        "token": "=",
        "inQuotes": false,
        "startIndex": 14,
        "endIndex": 14
    },
    {
        "token": "myProject",
        "inQuotes": false,
        "startIndex": 15,
        "endIndex": 20
    }
]
07:45:32.456 Expression tree: {
    "name": "AndExp",
    "subExpressions": [
        {
            "name": "PropertyComparisonExp",
            "propertyName": "isArchived",
            "operator": "=",
            "comparedValue": "false"
        },
        {
            "name": "AndExp",
            "subExpressions": [
                {
                    "name": "AttributeExistsExp",
                    "attributeType": "label",
                    "attributeName": "call",
                    "prefixMatch": false
                },
                {
                    "name": "LabelComparisonExp",
                    "attributeType": "label",
                    "attributeName": "project"
                }
            ]
        }
    ]
}
07:45:32.465 Slow 200 GET /api/search-note/teRiR0MOpOu6 took 12ms
```

Failing search =  = `#project=myProject #call`
```
07:45:01.698 Fulltext tokens: []
07:45:01.698 Expression tokens: [
    {
        "token": "#project",
        "inQuotes": false,
        "startIndex": 0,
        "endIndex": 7
    },
    {
        "token": "=",
        "inQuotes": false,
        "startIndex": 8,
        "endIndex": 8
    },
    {
        "token": "myProject",
        "inQuotes": false,
        "startIndex": 9,
        "endIndex": 14
    },
    {
        "token": "#call",
        "inQuotes": false,
        "startIndex": 16,
        "endIndex": 20
    }
]
07:45:01.698 Expression tree: {
    "name": "AndExp",
    "subExpressions": [
        {
            "name": "PropertyComparisonExp",
            "propertyName": "isArchived",
            "operator": "=",
            "comparedValue": "false"
        },
        {
            "name": "AndExp",
            "subExpressions": [
                {
                    "name": "LabelComparisonExp",
                    "attributeType": "label",
                    "attributeName": "project"
                },
                {
                    "name": "AttributeExistsExp",
                    "attributeType": "label",
                    "attributeName": "call",
                    "prefixMatch": false
                }
            ]
        }
    ]
}
07:45:01.705 200 GET /api/search-note/teRiR0MOpOu6 took 9ms
```

At first sight, I don't see the difference - the tokens & subexpressions change order, but that's it.

Another question, just to understand more about the search implementation: why doesn't the actual comparison value `myProject` appear in `LabelComparisonExp` subexpression? 

Hope this helps!

**Comment by zadam** [comment]
Yeah, I don't see a difference either.

> Another question, just to understand more about the search implementation: why doesn't the actual comparison value `myProject` appear in `LabelComparisonExp` subexpression?

Comparison value is in a dynamically created `comparator` function in the `LabelComparisonExp` object, but the simple dumping function ignores function objects and that's why it's not visible.

So far I don't have an idea of what could go on. As I understand the `myProject` value is not the real value with which you're testing. Does the real value contain some specific characters like dash, space, unicode or similar?

**Comment by wwwald** [comment]
Actual project name has a digit inside - but I tried with an `[a-z]+` project name and search behaves the same.

It must be something with my specific setup, in a plain Trilium setup (with a fresh empty `TRILIUM_DATA_DIR`) everything works as expected.
I don't have too much special things going on. I'm a heavy template user, have customized theming and have a single "Today" script running on startup, to provide a "Today" page shortcut.

Feel free to not waste your time on this - it could be a niche case, and I have my workaround.
